src/actions.ts

```diff
@@ -0,0 +1,708 @@
+import {
+    Offset,
+    ULabel,
+    ULabelAction,
+    ULabelActionRaw,
+    ULabelActionType,
+} from "..";
+import { FilterPointDistanceFromRow } from "./toolbox";
+import { AllowedToolboxItem } from "./configuration";
+import { filter_points_distance_from_line } from "./annotation_operators";
+import { NONSPATIAL_MODES } from "./annotation";
+
+import { log_message, LogLevel } from "./error_logging";
+
+// ================= Record Actions =================
+
+/**
+ * Record an action in the action stream to allow undo/redo functionality.
+ * Also trigger any listeners associated with the action.
+ *
+ * @param ulabel ULabel instance
+ * @param raw_action action to record
+ * @param is_redo whether ulabel action is a redo or not
+ * @param add_to_action_stream whether to add the action to the action stream
+ */
+export function record_action(ulabel: ULabel, raw_action: ULabelActionRaw, is_redo: boolean = false, add_to_action_stream: boolean = true) {
+    ulabel.set_saved(false);
+    const current_subtask = ulabel.get_current_subtask();
+    const annotation = current_subtask.annotations.access[raw_action.annotation_id];
+
+    // After a new action, you can no longer redo old actions
+    if (add_to_action_stream && !is_redo) {
+        current_subtask.actions.undone_stack = [];
+    }
+
+    // Stringify the undo/redo payloads
+    const act_type = raw_action.act_type;
+    const action: ULabelAction = {
+        act_type: act_type,
+        annotation_id: raw_action.annotation_id,
+        frame: raw_action.frame,
+        undo_payload: JSON.stringify(raw_action.undo_payload),
+        redo_payload: JSON.stringify(raw_action.redo_payload),
+        prev_timestamp: annotation.last_edited_at,
+        prev_user: annotation.last_edited_by,
+    };
+
+    // Add to stream
+    if (add_to_action_stream) {
+        current_subtask.actions.stream.push(action);
+
+        // Update annotation edit info
+        annotation.last_edited_at = ULabel.get_time();
+        annotation.last_edited_by = ulabel.config.username;
+    }
+
+    // Trigger any listeners for the action
+    on_start_annotation_spatial_modification(
+        ulabel,
+        action,
+    );
```

Comment:
It does feel slightly backwards to have the listeners/side effects always receive an action and then decide whether it matches - I think a lookup of side effects keyed by action types might make it easier to modify in the future. That would also provide a single spot to say "what happens when this action occurs (or is undone)", rather than "for what actions does this side effect occur". 

Let me know what you think.